src/server/csp.ts

```diff
@@ -0,0 +1,73 @@
+// ----- Imports ----- //
+
+import { createHash } from 'crypto';
+import { Design } from '@guardian/types/Format';
+
+import { BodyElement, ElementKind } from 'bodyElement';
+import { partition, Result } from 'types/result';
+import { Item } from 'item';
+import { compose } from 'lib';
+
+
+// ----- Types ----- //
+
+interface Assets {
+    scripts: string[];
+    styles: string[];
+}
+
+
+// ----- Functions ----- //
+
+const assetHash = (asset: string): string =>
+    createHash('sha256').update(asset).digest('base64');
+
+const extractInteractiveAssets = (elements: BodyElement[]): Assets =>
+    elements.reduce<Assets>(({ scripts, styles }, elem) => {
+        if (elem.kind === ElementKind.InteractiveAtom) {
+            return {
+                styles: [ ...styles, elem.css ],
+                scripts: elem.js.fmap(js => [ ...scripts, js ]).withDefault(scripts),
+            };
+        }
+
+        return { scripts, styles };
+    }, { scripts: [], styles: [] });
+
+const getElements = (item: Item): Result<string, BodyElement>[] =>
+    item.design === Design.Live ? item.blocks.flatMap(block => block.body) : item.body;
+
+const getValidElements = (item: Item): BodyElement[] =>
+    partition(getElements(item)).oks;
+
+const interactiveAssets = compose(extractInteractiveAssets, getValidElements);
+
+const assetHashes = (assets: string[]): string =>
+    assets.map(asset => `'sha256-${assetHash(asset)}'`).join(' ');
+
+const buildCsp = ({ styles, scripts }: Assets, twitter: boolean): string => `
+    default-src 'self';
+    style-src ${assetHashes(styles)} https://interactive.guim.co.uk ${twitter ? 'https://platform.twitter.com' : ''};
+    img-src 'self' https://i.guim.co.uk ${twitter ? 'https://platform.twitter.com https://syndication.twitter.com https://pbs.twimg.com data:' : ''};
+    script-src 'self' ${assetHashes(scripts)} https://interactive.guim.co.uk ${twitter ? 'https://platform.twitter.com https://cdn.syndication.twimg.com' : ''};
```

Comment:
dotcom are a bit more relaxed here, there's probably a reason for it 
`script-src https: 'unsafe-inline' 'unsafe-eval'`

Reply:
That could be commercial or other third parties though.

Reply:
Maybe, their client-side platform is more complex than ours.